src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+            {
+                if (statement is BlockSyntax block)
+                {
+                    return block.WithStatements(block.Statements.Insert(0, newStatement));
+                }
+                else
+                {
+                    return SyntaxFactory.Block(newStatement, statement);
+                }
+            }
+
+            private static BlockSyntax WrapWithBlock(StatementSyntax statement)
+                => statement is BlockSyntax block ? block : SyntaxFactory.Block(statement);
+
+            private static ExpressionSyntax ProcessCondition(ExpressionSyntax expression)
+            {
+                if (expression.Kind() == SyntaxKind.ParenthesizedExpression)
+                {
+                    expression = ((ParenthesizedExpressionSyntax)expression).Expression;
+                }
+
+                return expression.WithoutTrivia();
+            }
+
+            private static AnonymousObjectMemberDeclaratorSyntax CreateAnonymousObjectMemberDeclarator(string name)
+                => SyntaxFactory.AnonymousObjectMemberDeclarator(
+                    SyntaxFactory.NameEquals(SyntaxFactory.IdentifierName(name)), SyntaxFactory.IdentifierName(name));
+
+            private static StatementSyntax CreateJoinStatement(JoinClauseSyntax joinClause, StatementSyntax statement)
+            {
+                return SyntaxFactory.Block(
+                                SyntaxFactory.ForEachStatement(
+                                    joinClause.Type ?? VarNameIdentifier,
+                                    joinClause.Identifier,
+                                    joinClause.InExpression,
+                                    SyntaxFactory.Block(
+                                        SyntaxFactory.IfStatement(
+                                            SyntaxFactory.InvocationExpression(
+                                                SyntaxFactory.MemberAccessExpression(
+                                                    SyntaxKind.SimpleMemberAccessExpression,
+                                                    SyntaxFactory.PredefinedType(SyntaxFactory.Token(SyntaxKind.ObjectKeyword)),
+                                                    SyntaxFactory.IdentifierName(nameof(object.Equals))),
+                                                SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(
+                                                    new[] {
+                                                        SyntaxFactory.Argument(joinClause.LeftExpression),
+                                                        SyntaxFactory.Argument(joinClause.RightExpression.WithoutTrailingTrivia())}))),
+                                            statement)))).WithAdditionalAnnotations(Simplifier.Annotation);
+            }
+
+            private class ProcessingState
+            {
+                public ImmutableArray<string> VariableNames { get; private set; }
```

Comment:
>ImmutableArray<string> [](start = 23, length = 22)

Why use an immutable array for a list that is definitely going to be mutated? Consider using an array builder and simply adding names to it.